ksm_test.go

```diff
@@ -0,0 +1,341 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const ksmString = "ksmrules"
+const anonPagesMemory = 16777216 // Typically 4096 pages
+const run = "1"
+const interval = "10"
+const scan = "1000"
+
+func ksmTestPrepare() error {
+	newKSMRoot, err := ioutil.TempDir("", "cc-ksm-test")
+	if err != nil {
+		return err
+	}
+
+	defaultKSMRoot = newKSMRoot
+
+	memInfoFile, err := ioutil.TempFile("", "cc-ksm-meminfo")
+	if err != nil {
+		return err
+	}
+
+	memInfo = memInfoFile.Name()
+
+	_, err = memInfoFile.WriteString(fmt.Sprintf("AnonPages: %v kB", anonPagesMemory))
+	if err != nil {
+		return err
+	}
+
+	ksmTestRun, err := os.Create(filepath.Join(defaultKSMRoot, ksmRunFile))
+	if err != nil {
+		return err
+	}
+
+	ksmTestPagesToScan, err := os.Create(filepath.Join(defaultKSMRoot, ksmPagesToScan))
+	if err != nil {
+		return err
+	}
+
+	ksmTestSleepMillisec, err := os.Create(filepath.Join(defaultKSMRoot, ksmSleepMillisec))
+	if err != nil {
+		return err
+	}
+
+	defer ksmTestRun.Close()
+	defer ksmTestPagesToScan.Close()
+	defer ksmTestSleepMillisec.Close()
+
+	return nil
+}
+
+func ksmTestCleanup() {
+	os.RemoveAll(defaultKSMRoot)
+	os.RemoveAll(memInfo)
+}
+
+func TestKSMSysfsAttributeOpen(t *testing.T) {
+	pagesToScanSysFs := sysfsAttribute{
+		path: filepath.Join(defaultKSMRoot, ksmPagesToScan),
+	}
+
+	err := pagesToScanSysFs.open()
+	defer pagesToScanSysFs.close()
+
+	assert.Nil(t, err)
+}
+
+func TestKSMSysfsAttributeOpenNonExistent(t *testing.T) {
+	pagesToScanSysFs := sysfsAttribute{
+		path: filepath.Join(defaultKSMRoot, "foo"),
+	}
+
+	err := pagesToScanSysFs.open()
+	defer pagesToScanSysFs.close()
+
+	assert.NotNil(t, err)
+}
+
+func TestKSMSysfsAttributeReadWrite(t *testing.T) {
+	pagesToScanSysFs := sysfsAttribute{
+		path: filepath.Join(defaultKSMRoot, ksmPagesToScan),
+	}
+
+	err := pagesToScanSysFs.open()
+	defer pagesToScanSysFs.close()
+
+	assert.Nil(t, err)
```

Comment:
Note that you can simplify the code further using the trick we're starting to use on the runtime to avoid having to pass `t` to assert all the time:

```
assert := assert.New(t)

    :

assert.Nil(err)
```